sdks/python/apache_beam/runners/portability/fn_api_runner/worker_handlers.py

```diff
@@ -1087,14 +1150,49 @@ def append_raw(
   ):
     # type: (...) -> _Future
     with self._lock:
-      self._state[self._to_key(state_key)].append(data)
+      if state_key.WhichOneof('type') == 'ordered_list_user_state':
+        coder = TupleCoder(
+            [VarIntCoder(), coders.LengthPrefixCoder(BytesCoder())]).get_impl()
+
+        if self._to_key(state_key) not in self._ordered_list_keys:
+          self._ordered_list_keys[self._to_key(state_key)] = SortedSet()
+
+        for key, value in coder.decode_all(data):
+          persistent_state_key = beam_fn_api_pb2.StateKey()
+          persistent_state_key.CopyFrom(state_key)
+          persistent_state_key.ordered_list_user_state.range.start = key
+          persistent_state_key.ordered_list_user_state.range.end = key + 1
+          self._state[self._to_key(persistent_state_key)].append(
+              coder.encode((key, value)))
+          self._ordered_list_keys[self._to_key(state_key)].add(key)
+      else:
+        self._state[self._to_key(state_key)].append(data)
     return _Future.done()
 
   def clear(self, state_key):
     # type: (beam_fn_api_pb2.StateKey) -> _Future
     with self._lock:
       try:
-        del self._state[self._to_key(state_key)]
+        if state_key.WhichOneof('type') == 'ordered_list_user_state':
+          start = state_key.ordered_list_user_state.range.start
+          end = state_key.ordered_list_user_state.range.end
+          persistent_state_key = beam_fn_api_pb2.StateKey()
+          persistent_state_key.CopyFrom(state_key)
+          persistent_state_key.ordered_list_user_state.ClearField("range")
+          available_keys = self._ordered_list_keys[self._to_key(
+              persistent_state_key)]
+
+          keys_to_remove = []
+          for i in available_keys.irange(start, end, inclusive=(True, False)):
+            persistent_state_key.ordered_list_user_state.range.start = i
+            persistent_state_key.ordered_list_user_state.range.end = i + 1
+            del self._state[self._to_key(persistent_state_key)]
+            keys_to_remove.append(i)
```

Comment:
This is because in this loop, we are iterating the same SortedSet (aka. available_keys) using `irange()` to find out keys falling into a specific range. We cannot mutate the SortedSet, otherwise it will invalidate the iterator.

Alternatively, I can wrap  the `irange()` iterator with `list`. In this case, we can mutate the SortedSet in the same loop.